R/dependecies.R

```diff
@@ -0,0 +1,510 @@
+  if (length(previously_parsed_deps) == 0) return(0L)
+  previously_parsed_names <- vapply(previously_parsed_deps
+    , function(p) p[["package"]]$name, character(1))
+  if (package$name %in% previously_parsed_names) {
+    sel <- which(previously_parsed_names == package$name)
+    subsel <- vapply(previously_parsed_deps[sel]
+      , function(p) identical(p[["package"]]$version, package$version)
+      , logical(1))
+    if (any(subsel)) {
+      return(sel[subsel])
+    }
+  }
+  0L
+}
+
+#' Attach the latest available lockbox version to a package
+add_latest_version_in_lockbox <- function(package) {
+  package$latest_version_in_lockbox <- max_package_version(
+    list.files(file.path(lockbox_library(), package)))
+  package
+}
+
+#' Take a vector of versions and find the max without coercing to package_version
+max_package_version <- function(versions) {
+  if (length(versions) == 0) return(NULL)
+  formatted_versions <- package_version(as.character(versions))
+  versions[which(formatted_versions == max(formatted_versions))[1]]
+}
+
+#' Check a dependency list for inclusion in the lockfile and replace the package
+#' with the locked version if it does appear there.  Also throw
+#' an error if we require a dependency version greater than that specified by
+#' the lockfile.
+replace_with_lock <- function(package, lock) {
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- lock[[which(lock_names == package$name)[1]]]
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
+      package_version(as.character(locked_package$version))) {
+        stop(paste0("Dependency: \'", package$name, ", Version: ", package$version
+          , " is required, but lockbox is locked at version: "
+          , as.character(locked_package$version)))
+    }
+    package <- locked_package
+  } else {
+    package$is_dependency_package <- TRUE
+  }
+  if (!"remote" %in% names(package) || is.na(package$remote)) {
+    package$remote <- "CRAN"
+  }
+  package <- as.locked_package(package)
+  if (package$is_dependency_package && !package$name %in%
+    as.character(installed.packages(priority = "base")[,1])) {
+      package$latest_version <- package$latest_version_in_lockbox %||%
+        get_latest_version(package)
+  }
+  package
+}
+
+get_latest_version <- function(package) {
+  if (package$remote == "CRAN") {
+    get_available_cran_version(package)
+  } else {
+    version_from_remote(package)
+  }
+}
+
+#' Combine two lists of dependencies via version comparisons.  Keep packages
+#' found in list1 on the left side of the entirety of list2, while moving
+#' the parent package to the space after it's rightmost dependency found in list2.
+combine_dependencies <- function(list1, list2, current_parent) {
+  names1 <- vapply(list1, function(obj) obj$name, character(1))
+  names2 <- vapply(list2, function(obj) obj$name, character(1))
```

Comment:
Two suggestions:

1.) Use `vapply(list1, `[[`, character(1), "name")`.

2.) DRY up by just processing as a list:

```R
names <- lapply(list(list1, list2), function(lst) vapply(lst, `[[`, character(1), "name"))
```

and then use `names[[1]]` and `names[[2]]` instead of `names1` and `names2`.